lib/lz4.c

```diff
@@ -602,8 +640,19 @@ LZ4_FORCE_INLINE int LZ4_compress_generic(
                 if (unlikely(forwardIp > mflimit)) goto _last_literals;
 
                 match = LZ4_getPositionOnHash(h, cctx->hashTable, tableType, base);
-                if (dict==usingExtDict) {
+                if (dictDirective == usingExtDictCtx) {
                     if (match < (const BYTE*)source) {
+                        /* there was no match, try the dictionary */
+                        /* TODO: use precalc-ed hash? */
+                        match = LZ4_getPosition(ip, dictCtx->hashTable, byU32, dictBase);
+                        refDelta = dictDelta;
+                        lowLimit = dictionary;
+                    } else {
+                        refDelta = 0;
+                        lowLimit = (const BYTE*)source;
+                    }
+                } else if (dictDirective==usingExtDict) {
+                    if (match < (const BYTE*)source && dictionary != NULL) {
```

Comment:
Is it possible to have both `(dictDirective==usingExtDict)` and `dictionary == NULL` ?
If not, this condition could be replaced by an `assert()`, in order to be removed from optimised binary (this is the hot loop, every branch matters).

Reply:
Apparently it happens, though I don't know when/why.

Reply:
This should be analysed then.
If this combination can happen, it should be a legitimate construction.
There might be an optimisation opportunity left there.